### DISCUSSION TITLE
How do I build an array using MPh

### DISCUSSION THREAD

**Discussion by aboschTDA** [MPh-py/MPh#133] [General]
To javascript for a given array seems to follow the format:
model.component("comp1").geom("geom1").create("arr1", "Array");
model.component("comp1").geom("geom1").feature("arr1").active(False);
model.component("comp1").geom("geom1").feature("arr1").set("fullsize", ["1", "S_int"]);
model.component("comp1").geom("geom1").feature("arr1").set("displ", ["0", "-S_s"]);
model.component("comp1").geom("geom1").feature("arr1").selection("input").set("r2");

I cant seem to figure out the MPh equivalent for:
model.component("comp1").geom("geom1").feature("arr2").selection("input").set("r2");

I had assumed the MPh equivalent code would be:
array_1 = geometry.create('Array', name = 'array_1')
array_1.set('input', 'thing to array')
array_1.property('fullsize', ['1', 'S_int'])
array_1.property('displ', ['0', 'S_s'])

or array_1.property('input', 'thing to array') 

Neither seem to work. Maybe I am missing something obvious?

**Comment by john-hen** [comment]
I think your question is about "selecting a geometry object as an input feature", rather than building the array. The array (of geometry objects) should build just fine.

I don't know what you mean by "to javascript". As far as I can tell, the code snippet you posted is Python. In other words, *not* Java code. And certainly not "Javascript", which is something else entirely.

Usually, I have `model` refer to an `mph.Model` object in Python code. And then access the "Java-like" API via `model.java`.
```python
model.java.component("comp1").geom("geom1").create("arr1", "Array");
model.java.component("comp1").geom("geom1").feature("arr1").active(False);
model.java.component("comp1").geom("geom1").feature("arr1").set("fullsize", ["1", "S_int"]);
model.java.component("comp1").geom("geom1").feature("arr1").set("displ", ["0", "-S_s"]);
model.java.component("comp1").geom("geom1").feature("arr1").selection("input").set("r2");
```

It looks like you assigned `model` to the `.java` property of the `Model` object in your code snippet. But then it should work. Does it not? Please explain what works and what doesn't. I can't run your code to find out: it's incomplete. As far as I can tell, what you have there does work "using MPh", i.e. in Python.

Maybe you're looking for a way to make the Python code more "pythonic" (i.e., less "Java-like"). Then you'd want to call [`select()`] on the node that needs the input. Unfortunately, that won't work for geometry features, as the documentation states:
> Raises `NotImplementedError` if the node (that this method is called on) is a geometry node. These may be supported in a future release. Meanwhile, access their Java methods directly. 

So *if* that's what you want, it's simply not implemented. But, as the documentation also notes, you can just do... what you're already doing, right?

[`select()`]: https://mph.readthedocs.io/en/1.2/api/mph.Node.html#mph.Node.select

**Reply by aboschTDA** [reply]
Hi John, 

Thank you for responding. My apologies for the confusion... my issue turned out to be a dumb oversight on my part. I had built my initial geometry using the python like syntax and I was referencing the python variable name in the java like array code instead of its tag. The code snippet I had provided was from a COMSOL javafile export and still had a javalike tag which is why it looked like it would run. 

My working code is thus: 

import mph
client = mph.start()
model = client.create('Test')

parameters = model/'parameters'
(parameters/'Parameters 1').rename('parameters')

model.parameter('CH', '1[in]')
model.description('CH', 'Cavity Height')
model.parameter('CW', '1[in]')
model.description('CW', 'Cavity Width')
model.parameter('S_s', '1[mm]')
model.description('S_s', 'Spacing')
model.parameter('S_c', '5')
model.description('S_c', 'Sorbent Stack Count')


components = model/'components'
components.create(True, name = 'component')

geometries = model/'geometries'
geometry = geometries.create(2, name = 'geometry')

cavity = geometry.create('Rectangle', name = 'cavity')
cavity.property('pos', ['0','0'])
cavity.property('base', 'center')
cavity.property('size', ['CW', 'CH'])

cavity_tag = cavity.tag()

model.java.component("comp1").geom("geom1").create("arr1", "Array");
model.java.component("comp1").geom("geom1").feature("arr1").active(True);
model.java.component("comp1").geom("geom1").feature("arr1").set("fullsize", ["1", "S_s"]);
model.java.component("comp1").geom("geom1").feature("arr1").set("displ",["0", "S_s"]);
model.java.component("comp1").geom("geom1").feature("arr1").selection("input").set(cavity_tag)

model.save('test.mph')


Thanks!
